### PR TITLE
Andy/auto increment

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

File: sql/numbertype.go [hunk 1]
```diff
@@ -86,6 +86,70 @@ func MustCreateNumberType(baseType query.Type) NumberType {
 	return nt
 }
 
+func NumericZeroValue(t Type) interface{} {
```
Comment (zachmu) [inline sql/numbertype.go:89]: Any reason you can't use Type.Zero()? This is just a subset of that, for the numeric types.

File: sql/numbertype.go [hunk 2]
```diff
@@ -86,6 +86,70 @@ func MustCreateNumberType(baseType query.Type) NumberType {
...
+	case sqltypes.Float32:
+		return float32(0)
+	case sqltypes.Float64:
+		return float64(0)
+	default:
+		panic(fmt.Sprintf("%v is not a valid number base type", nt.baseType.String()))
+	}
+}
+
+func NumericUnaryValue(t Type) interface{} {
```
Comment (zachmu) [inline sql/numbertype.go:121]: Just get the zero and add one, no need for this method

File: sql/parse/parse.go [hunk 3]
```diff
@@ -41,6 +41,8 @@ var (
 
 	ErrUnknownIndexColumn = errors.NewKind("unknown column: '%s' in %s index '%s'")
 
+	ErrInvalidAutoIncCols = errors.NewKind("there can be only one auto column and it must be defined as a key")
```
Comment (zachmu) [inline sql/parse/parse.go:44]: auto_increment

**Review by zachmu — COMMENTED** [review]
Needs a little work before checking in. Two issues:

1) Semantics for auto increment when the user supplies a value are wrong. For 0 or null, use the auto increment. Otherwise, honor the value given.
2) Analysis for the above should happen in the analyzer, not at runtime. i.e. the analyzer needs to decide whether to use an auto increment or the supplied value

File: sql/numbertype.go (hunk 2 above)
Comment (andrew-wm-arthur) [reply sql/numbertype.go:121]: `Type.Zero() + 1` is a type mismatch

File: enginetest/insert_queries.go [hunk 4]
```diff
@@ -328,6 +328,78 @@ var InsertQueries = []WriteQueryTest{
...
+		"SELECT * FROM auto_increment_tbl ORDER BY pk",
+		[]sql.Row{
+			{1, 11},
+			{2, 22},
+			{3, 33},
+			{4, 44},
+		},
+	},
+	{
+		"INSERT INTO auto_increment_tbl values (5, 44)",
```
Comment (zachmu) [inline enginetest/insert_queries.go:377]: Should have another test for inserting a value lower than the current max value (need sparse values in the table for this)

File: sql/expression/autoincrement.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+	return &AutoIncrement{
+		BinaryExpression{Left: lastInsertId, Right: given},
+		nil,
+		sync.Once{},
+	}, nil
+}
+
+// Resolved implements the Expression interface.
+func (i *AutoIncrement) Resolved() bool {
```
Comment (zachmu) [inline sql/expression/autoincrement.go:33]: Think you can omit this, handled by BinaryExpression

**Review by zachmu — APPROVED** [review]

**Pull request merged**